include/nbl/builtin/hlsl/shapes/beziers.hlsl

```diff
@@ -133,610 +198,205 @@ namespace shapes
-            bc0 /= bc6;
-            
-            //compute derivatives of this polynomial
-
-            float b0 = bc1 / 6.;
-            float b1 = 2. * bc2 / 6.;
-            float b2 = 3. * bc3 / 6.;
-            float b3 = 4. * bc4 / 6.;
-            float b4 = 5. * bc5 / 6.;
-
-            float4 c1 = float4(b1,2.*b2,3.*b3,4.*b4)/5.;
-            float3 c2 = float3(c1[1],2.*c1[2],3.*c1[3])/4.;
-            float2 c3 = float2(c2[1],2.*c2[2])/3.;
-            float c4 = c3[1]/2.;
-
-            float4 roots_drv = float4(1e38,1e38,1e38,1e38);
-
-            int num_roots_drv = solve_quartic(c1,roots_drv);
-            sort_roots4(roots_drv);
-
-            float ub = upper_bound_lagrange5(b0,b1,b2,b3,b4);
-            float lb = lower_bound_lagrange5(b0,b1,b2,b3,b4);
-
-            float3 a = float3(1e38,1e38,1e38);
-            float3 b = float3(1e38,1e38,1e38);
-
-            float3 roots = float3(1e38,1e38,1e38);
-
-            int num_roots = 0;
-            
-            //compute root isolating intervals by roots of derivative and outer root bounds
-            //only roots going form - to + considered, because only those result in a minimum
-            if(num_roots_drv==4)
+            else
             {
-                if(eval_poly5(b0,b1,b2,b3,b4,roots_drv[0]) > 0.)
+                float_t z = sqrt(-p);
+                float_t v = acos( q/(p*z*2.0) ) / 3.0;
+                float_t m = cos(v);
+                float_t n = sin(v)*1.732050808;
+                //float3_t t = float3_t(m + m, -n - m, n - m) * z - kx;
+                
+                float2_t t[3];
+                t[0] = (m + m) * z - kx;
+                t[1] = (-n - m) * z - kx;
+                t[2] = (n - m) * z - kx;
+                
+                for(uint32_t i = 0u; i < 3u; i++)
                 {
-                    a[0]=lb;
-                    b[0]=roots_drv[0];
-                    num_roots=1;
+                    t[i] = clamp(t[i], 0.0, 1.0);
+                    t[i] = clipper(t[i].x);
                 }
-
-                if(sign(eval_poly5(b0,b1,b2,b3,b4,roots_drv[1])) != sign(eval_poly5(b0,b1,b2,b3,b4,roots_drv[2])))
+                
+                // 3 roots
+                float2_t qos = CsubPos + (B + A*t[0].x)*t[0].x;
+                float_t dis = dot(qos,qos);
+                
+                res = float2_t(dis,t[0].x);
+                
+                if(t[0].x != t[0].y)
                 {
-                    if(num_roots == 0){
-                        a[0]=roots_drv[1];
-                        b[0]=roots_drv[2];
-                        num_roots=1;
-                    }
-                    else{
-                        a[1]=roots_drv[1];
-                        b[1]=roots_drv[2];
-                        num_roots=2;
-                    }
+                    qos = CsubPos + (B + A*t[0].y)*t[0].y;
+                    dis = length(qos);
+                    if(dis < res.x) res = float2_t(dis, t[0].y);

```

Comment:
instead of doing this 3 times, just move it to the for loop above :)
there you can also completely ignore both **clipped t values** if the distance is larger than some max distance

1. find the `qos` for the t[i].x value before before clipping, and set the current dist to that (res.x)
2. If it's larger than a (max dist ^ 2) then `continue` to the next iteration
3. If not proceed with the clipping and if it was actually clipped find new `qos`(s) and do what you're doing now

Reply:
done